### PR TITLE
Adding pyroscope and memray dependencies

### PR DESCRIPTION
**Pull request by vishnuchalla**
## Description

Adding pyroscope and memray dependencies

## Type of change

- [ ] Refactor
- [x] New feature
- [ ] Bug fix
- [ ] CVE fix
- [ ] Optimization
- [ ] Documentation Update
- [x] Configuration Update
- [x] Bump-up dependent library
- [ ] Bump-up library or tool used for development (does not change the final image)

## Closes
* https://issues.redhat.com/browse/OLS-964
* https://issues.redhat.com/browse/OLS-965

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] PR has passed all pre-merge test jobs.
- [ ] If it is a core feature, I have added thorough tests.

## Testing
Tested and verified in local. Both pyroscope and memray get triggered as a part of OLS's startup activity.

### CODE REVIEW

**Review by tisnik — COMMENTED** [review]
would it be possible to add these dependencies only for "devel mode"? The same way as GradIO. It will keep number of packages on lowest level possible + images will be smaller + (most important) less CVE-prone

**Comment by vishnuchalla** [comment]
> would it be possible to add these dependencies only for "devel mode"? The same way as GradIO. It will keep number of packages on lowest level possible + images will be smaller + (most important) less CVE-prone

We will need these dependencies when OLS is running through the operator so that we can capture profiles while load testing. Move them to them "dev' group will make it impossible.

**Comment by tisnik** [comment]
> > would it be possible to add these dependencies only for "devel mode"? The same way as GradIO. It will keep number of packages on lowest level possible + images will be smaller + (most important) less CVE-prone
> 
> We will need these dependencies when OLS is running through the operator so that we can capture profiles while load testing. Moving them to them `dev` group will make it impossible.

ok, but at least loading should be performed if really needed (and configured in config.yaml) right?

**Comment by vishnuchalla** [comment]
> > > would it be possible to add these dependencies only for "devel mode"? The same way as GradIO. It will keep number of packages on lowest level possible + images will be smaller + (most important) less CVE-prone
> > 
> > 
> > We will need these dependencies when OLS is running through the operator so that we can capture profiles while load testing. Moving them to them `dev` group will make it impossible.
> 
> ok, but at least loading should be performed if really needed (and configured in config.yaml) right?

When we say loading, are you referring to `import` statement in specific?

**Comment by tisnik** [comment]
> > > > would it be possible to add these dependencies only for "devel mode"? The same way as GradIO. It will keep number of packages on lowest level possible + images will be smaller + (most important) less CVE-prone
> > > 
> > > 
> > > We will need these dependencies when OLS is running through the operator so that we can capture profiles while load testing. Moving them to them `dev` group will make it impossible.
> > 
> > 
> > ok, but at least loading should be performed if really needed (and configured in config.yaml) right?
> 
> When we say loading, are you referring to `import` statement in specific?

Yes, in the similar way we "optimized" Gradio packages loading

**Comment by vishnuchalla** [comment]
> > > > would it be possible to add these dependencies only for "devel mode"? The same way as GradIO. It will keep number of packages on lowest level possible + images will be smaller + (most important) less CVE-prone
> > > 
> > > 
> > > We will need these dependencies when OLS is running through the operator so that we can capture profiles while load testing. Moving them to them `dev` group will make it impossible.
> > 
> > 
> > ok, but at least loading should be performed if really needed (and configured in config.yaml) right?
> 
> When we say loading, are you referring to `import` statement in specific?

I just conducted some experiments to re-evaluate your suggestion. I think moving them to `dev` makes sense. 
And profiling under load doesn't add any value because of huge amount of data that it spits and is impossible to interpret. Example memory profile from `memray`
```
-rw-r--r--. 1 1000700000 root       5.2G Sep 11 02:59 output.bin
```
And pyroscope also captured information that is impossible to interpret. So In order to get better understand on profiling, keeping them in `dev` group and capturing profiles against one single profile is sufficient enough to take a look at things in detail.
cc: @syedriko @xrajesh

File: Makefile [hunk 1]
```diff
@@ -48,6 +48,16 @@ update-deps: ## Check pyproject.toml for changes, update the lock file if needed
 run: ## Run the service locally
 	python runner.py
 
+memray-run: ## Run the service locally using memray
+	python -m memray run -o ./memory_profile.bin runner.py
+
+memray-flamegraph: ## Extract flamegraph from memray output
```
Comment (tisnik) [inline Makefile:54]: you can make it as dependency for memray-run, right

File: runner.py [hunk 2]
```diff
@@ -148,10 +149,47 @@ def start_uvicorn():
...
-    # create and start the rag_index_thread - allows loading index in
-    # parallel with starting the Uvicorn server
-    rag_index_thread = threading.Thread(target=load_index)
-    rag_index_thread.start()
+    if config.dev_config.pyroscope_url:
+        try:
+            import requests
+
+            response = requests.get(config.dev_config.pyroscope_url, timeout=60)
+            if response.status_code == 200:
```
Comment (tisnik) [inline runner.py:157]: requests.codes.ok

File: Makefile (hunk 1 above)
Comment (vishnuchalla) [reply Makefile]: Changed as suggested.

**Comment by vishnuchalla** [comment]
@tisnik can you take a look on this one as well please?

**Comment by vishnuchalla** [comment]
> /hold

@tisnik Is there a reason to hold this PR?

**Comment by vishnuchalla** [comment]
@tisnik @xrajesh @syedriko since we are post GA, can we unhold and take this forward?

**Comment by vishnuchalla** [comment]
@xrajesh @tisnik 👋🏼  PTAL once you get some time. Thank you

File: pyproject.toml [hunk 3]
```diff
@@ -56,6 +56,8 @@ dev = [
     "types-requests==2.32.0.20240622",
     "gradio==4.44.1",
     "boto3==1.34.145",
+    "pyroscope-io>=0.8.7",
```
Comment (tisnik) [inline pyproject.toml:59]: please lock version for both new packages

File: pyproject.toml [hunk 4]
```diff
@@ -56,6 +56,8 @@ dev = [
...
+    "memray>=1.14.0",
```
Comment (tisnik) [inline pyproject.toml:60]: dtto

File: pdm.lock [hunk 5]
```diff
@@ -4,8 +4,11 @@
 [metadata]
```
Comment (tisnik) [inline pdm.lock:4]: was it generated on MacOS? seems a bit weird that so many changes were incorporated

File: runner.py [hunk 6]
```diff
@@ -60,6 +59,43 @@ def load_index():
     # init loading of query redactor
     config.query_redactor
 
+    if config.dev_config.pyroscope_url:
```
Comment (tisnik) [inline runner.py:62]: could you please move it somewhere into `ols/src/utils/pyroscope` or so? TYVM!

**Review by tisnik — CHANGES_REQUESTED** [review]
Seems fine in overall, just have few comments/nitpicks. Please look at them

File: pyproject.toml (hunk 3 above)
Comment (vishnuchalla) [reply pyproject.toml]: Done

File: pyproject.toml (hunk 4 above)
Comment (vishnuchalla) [reply pyproject.toml]: locked as suggested.

File: pdm.lock (hunk 5 above)
Comment (vishnuchalla) [reply pdm.lock:4]: I used a fedora linux machine.

File: runner.py (hunk 6 above)
Comment (vishnuchalla) [reply runner.py:63]: Moved it to `ols/utils/pyroscope` as suggested.

**Comment by vishnuchalla** [comment]
@xrajesh @tisnik looks like good to go. Went through all the CI tests

**Comment by tisnik** [comment]
/lgtm
/approve
/override "Red Hat Konflux / ols-enterprise-contract / lightspeed-service"

**Pull request merged**